Inlining checkout & Single receiver mode

mikulas-mrva opened the pull request:
Fixes multiple open issues

  * Implementing single receiver mode
  * Inlining Web3checkout
  * CI timeout in pip install !207
  * Adding Goerli testnet

mikulas-mrva committed:
Adding web3modal as a build package for now, changing multiple receiver mode to single receiver, deleting check for available receiving addresses

mikulas-mrva committed:
removing wallet upload models and views, adding wallet connection js, transaction details view in progress

mikulas-mrva committed:
Added SignedMessage model, view and serializer to send payment details to the FE

mikulas-mrva committed:
Finished the FE code outline, added Permissions to allow POST and GET requests for anonymous users

mikulas-mrva committed:
Switching to the correect way of signing a message with Metamask

mikulas-mrva committed:
Changing data signing to typed data

mikulas-mrva committed:
Fixing data signing

mikulas-mrva committed:
Confirming payments, deleting unused permissions module

mikulas-mrva committed:
Improving user flow of the payment process, displaying a message when the user should sign a message and confirm a transaction

mikulas-mrva committed:
Missing migration

mikulas-mrva committed:
Changing raw_message to text field to circumvent field length limitations

mikulas-mrva committed:
Moving retry triggers from inline to js, adding transaction confirmation text for the user.

mikulas-mrva committed:
Handling user dialogue cancellations gracefully.

mikulas-mrva committed:
Do not hide success message if the transaction was submitted successfully

mikulas-mrva committed:
chain_id should be larger than uint8 to accomodate all chains

mikulas-mrva committed:
Fixing a typo, using signTypedData_v4 for signing

mikulas-mrva committed:
Network name should not be uppercase in _RPC_URL

mikulas-mrva committed:
Extending chain id field

mikulas-mrva committed:
Bug fixes

mikulas-mrva committed:
Don't let the user pay for multiple order payments wwithin one order

mikulas-mrva committed:
Disabling refunds

mikulas-mrva committed:
fixes:
* Adding ERC20 ABI view for ERC20 payments
* making sure only one order_payment can be paid for
* no longer disabling the "Connect Wallet and pay" button

mikulas-mrva committed:
Confirming ERC20 payments

mikulas-mrva committed:
Don't confirm payments based on blocks that are too young

mikulas-mrva committed:
Display transaction hash on refresh, display transaction details if there's no hash yet.

mikulas-mrva committed:
Add ERC20 balance check to prevent most failing transactions to be submitted

mikulas-mrva committed:
Automatically confirming payments, ignoring payments that didn't go through

mikulas-mrva committed:
Preventing a payment to go through in case of insufficient balance error, reloading the page on account change instead of triggering the logic immediately

mikulas-mrva committed:
Only display "please sign a message dialogue" in case it's about to pop up in the browser.

mikulas-mrva committed:
Only display "please confirm a transaction" right before a transaction dialog is about to pop up

mikulas-mrva committed:
Optimizing "order_accepting_payments" criteria into a single db query

rahul-kothari commented on pretix_eth/exporter.py:
```diff
@@ -57,7 +56,7 @@ def refund_to_row(refund):
     token_amount = refund.info_data.get("amount", "")
 
-    wallet_address = WalletAddress.objects.filter(order_payment=refund.payment).first()
+    wallet_address = None  # todo WalletAddress.objects.filter(order_payment=refund.payment).first()
```
out of the loop - why are we not showing wallet addresses anymore?

mikulas-mrva replied:
We have deleted the model altogether for single receiver mode, now it's just a settings field. This was a temporary hack -- commented it out to prevent it from crashing due to the missing model. Fixed now :)

mikulas-mrva replied:
Deleted the refund logic now, the single receiver mode doesn't support refunds for privacy reasons.

rahul-kothari commented on pretix_eth/management/commands/confirm_payments.py:
```diff
@@ -43,54 +47,97 @@ def handle(self, *args, **options):
+                    try:
+                        transaction_details = w3.eth.getTransaction(signed_message.transaction_hash)
+                    except TransactionNotFound:
+                        if signed_message.age > 30*60:
```
can this be a config? The amount of time before we invalidate a transaction?

mikulas-mrva replied:
It's a config now! This is actually the amount of time before a customer can try to pay again in case their first payment didn't confirm.

rahul-kothari commented on pretix_eth/management/commands/confirm_payments.py:
```diff
@@ -43,54 +47,97 @@ def handle(self, *args, **options):
+                # native asset
+                if token.IS_NATIVE_ASSET:
+                    try:
+                        transaction_details = w3.eth.getTransaction(signed_message.transaction_hash)
```
why is this not getTransactionReceipt?

rahul-kothari commented on pretix_eth/management/commands/confirm_payments.py:
```diff
@@ -6,18 +6,21 @@
 logger = logging.getLogger(__name__)
 
 
+SAFETY_BLOCK_COUNT = 5
```
can this be a config value pleasE?

mikulas-mrva replied:
Good point, it's a config now, thanks!

rahul-kothari commented on pretix_eth/models.py:
```diff
@@ -1,70 +1,43 @@
-class WalletAddressQuerySet(models.QuerySet):
-    def for_event(self, event: Event) -> models.QuerySet:
-        return self.filter(event=event)
+    MAX_AGE_UNCONFIRMED = 30 * 60  # allow retry payment after 30mins
```
we have replicated this at many places. Let's move it to a config.yaml or something?

mikulas-mrva replied:
This was class constant actually not used in the end, and it's been moved to settings.

rahul-kothari commented on pretix_eth/management/commands/confirm_refunds.py:
```diff
@@ -13,9 +13,6 @@
 
-from pretix_eth.models import (
-    WalletAddress,
-)
```
how is confirm refunds working if we have removed this import even though the file still uses WalletAddress?

mikulas-mrva replied:
It wasn't finished yet. There are no refunds now due to private key concerns, so refunds are handled manually

mikulas-mrva committed:
Fixing AttributeError: Manager isn't accessible via SignedMessage instances

mikulas-mrva committed:
Using getReceipt in both ETH and DAI payment confirmation

mikulas-mrva committed:
Using getReceipt in both ETH and DAI payment confirmation

mikulas-mrva committed:
refactoring confirm_payments to use reeceieptsto check transaction status

mikulas-mrva committed:
Refactoring js, splitting into multiple files

mikulas-mrva committed:
Refactoring js, splitting into multiple files, fixes

mikulas-mrva committed:
Periodic payment status change check should be running even if user refreshes the page while the payment is panding.

So far it was triggered by JS upon transaction hash submission to the BE, now it runs on page load also when needed.

mikulas-mrva committed:
bugfixes

mikulas-mrva committed:
Deleting confirm_refunds.py management command as there are no refunds in single receiver mode

mikulas-mrva committed:
Adding an endpoint that can access the status of an Order without auth restrictions that are in place on the Pretix API endpoints we were using previously.
Also deleting general error callbacks that were resetting the error message, covering useful error messages with generic ones.

mikulas-mrva committed:
Updating Web3.js to a new version 1.7.5

mikulas-mrva committed:
Updating Web3Modal to a hotfixed version branched out from latest version, but including our fix of Firefox CSP Metamask issue https://github.com/MetaMask/metamask-extension/issues/3133

mikulas-mrva committed:
Adding *.infura.io to CSP

mikulas-mrva committed:
Fixing some CSP issues, adding signed message data to Order detail control view

mikulas-mrva committed:
Updating CSP and fixing javascript issues that require a double click, removing WalletConnect

mikulas-mrva committed:
Handling missing providers gracefully, deleting a ton of debuggers;

mikulas-mrva committed:
Bugfix: retrieving web3modal correctly

mikulas-mrva committed:
Bugfix: retrieving web3modal correctly

mikulas-mrva committed:
bugfixes

mikulas-mrva committed:
bugfixes - two different kinds of provider

mikulas-mrva committed:
adding loggers to confirm_payment mgmt commands

mikulas-mrva committed:
adding loggers to confirm_payment mgmt commands, log level changed to info, but managed by verbosity argument

mikulas-mrva committed:
Also checking for cancelled OrderPayments

mikulas-mrva committed:
ERC20ABIView should be public

mikulas-mrva committed:
Confirming DAI payments using the correct received address

mikulas-mrva committed:
Checking ERC20 contract address as well

mikulas-mrva committed:
Changing `url` template tag to `eventurl event`

mikulas-mrva committed:
Changing `url` template tag to `eventurl event`

mikulas-mrva committed:
Views should be compatible with multidomain as well as signel event pretix deployments

mikulas-mrva committed:
Arbitrum DAI should have a checksum address

mikulas-mrva committed:
All DAI addresses should be checksum addresses

rahul-kothari commented on pretix_eth/exporter.py:
```diff
@@ -57,7 +56,7 @@ def refund_to_row(refund):
     token_amount = refund.info_data.get("amount", "")
 
-    wallet_address = WalletAddress.objects.filter(order_payment=refund.payment).first()
+    wallet_address = None  # todo WalletAddress.objects.filter(order_payment=refund.payment).first()
```
What is the point of this and the next line if we are setting it to None?

mikulas-mrva replied:
Refunds are disabled under single receiver node for private key management reasons.

mikulas-mrva committed:
* code cleanup
* displaying block explorer links on the FE
* displaying a "please change your network to X" prompt
* reformatting control dashboard payment info

mikulas-mrva committed:
adding links to control.html list of payments

mikulas-mrva committed:
Adding payment data to exports

mikulas-mrva committed:
Fixing tests

mikulas-mrva committed:
Linter complaints

mikulas-mrva committed:
Linter complaints

mikulas-mrva committed:
Removing Refunds from the exports logic, we don't support them anymore

mikulas-mrva committed:
Moving SAFETY_BLOCK_COUNT and PAYMENT_NOT_RECIEVED_RETRY_TIMEOUT values into settings to make them configurable

mikulas-mrva committed:
Added a json validator for TOKEN_RATES field

mikulas-mrva committed:
Removing an unused constant

mikulas-mrva committed:
Fixing a template bug

mikulas-mrva committed:
Rinkeby in Peace

mikulas-mrva committed:
safety_block_count can be None

mikulas-mrva committed:
Using Correct DAI contract

mikulas-mrva committed:
Adding Sepolia ETH to the config, not sure what the correct DAI contract address is currently, to be be added later.

mikulas-mrva committed:
Putting Rinkeby back in place for Exports, added a DISABLED flag to prevent it from being used in new payments

mikulas-mrva committed:
Adding Token Rate at time of order to ETH payment export

mikulas-mrva committed:
Export token price in ether, not wei

mikulas-mrva committed:
Trying to fix infinite pip backtracking and adding python 3.10, 3.11 to GitHub workflows

mikulas-mrva committed:
Fixinf dependency conflicts for Py<3.8

mikulas-mrva committed:
Temporarily disabling pip check

mikulas-mrva committed:
Fixing tests, pinning web3 to a beta of v6 temporarily to resolve an issue with protobuf
